addon-resizer/nanny/nanny_lib.go

```diff
@@ -74,41 +90,73 @@ type ResourceEstimator interface {
+
 	for i := 0; true; i++ {
 		if i != 0 {
 			// Sleep for the poll period.
 			time.Sleep(pollPeriod)
 		}
 
-		// Query the apiserver for the number of nodes.
-		num, err := k8s.CountNodes()
-		if err != nil {
-			log.Error(err)
-			continue
+		if updateResources(k8s, est, time.Now(), lastChange, scaleDownDelay, scaleUpDelay, threshold) {
+			lastChange = time.Now()
 		}
-		log.V(4).Infof("The number of nodes is %d", num)
+	}
+}
 
-		// Query the apiserver for this pod's information.
-		resources, err := k8s.ContainerResources()
-		if err != nil {
-			log.Errorf("Error while querying apiserver for resources: %v", err)
-			continue
-		}
+// updateResources counts the number of nodes, estimates the expected
+// ResourceRequirements, compares them to the actual ResourceRequirements, and
+// updates the deployment with the expected ResourceRequirements if necessary.
+// It returns true if deployment has been updated.
+func updateResources(k8s KubernetesClient, est ResourceEstimator, now, lastChange time.Time, scaleDownDelay, scaleUpDelay time.Duration, threshold uint64) bool {
+	// Query the apiserver for the number of nodes.
+	num, err := k8s.CountNodes()
+	if err != nil {
+		log.Error(err)
+		return false
+	}
+	log.V(4).Infof("The number of nodes is %d", num)
 
-		// Get the expected resource limits.
-		expResources := est.scaleWithNodes(num)
+	// Query the apiserver for this pod's information.
+	resources, err := k8s.ContainerResources()
+	if err != nil {
+		log.Errorf("Error while querying apiserver for resources: %v", err)
+		return false
+	}
 
-		// If there's a difference, go ahead and set the new values.
-		if !shouldOverwriteResources(int64(threshold), resources.Limits, resources.Requests, expResources.Limits, expResources.Requests) {
-			log.V(4).Infof("Resources are within the expected limits. Actual: %+v Expected: %+v", *resources, *expResources)
-			continue
+	// Get the expected resource limits.
+	expResources := est.scaleWithNodes(num)
+
+	// If there's a difference, go ahead and set the new values.
+	overwrite, op := shouldOverwriteResources(int64(threshold), resources.Limits, resources.Requests, expResources.Limits, expResources.Requests)
+	if !overwrite {
+		if log.V(4) {
+			log.V(4).Infof("Resources are within the expected limits. Actual: %+v Expected: %+v", jsonOrValue(*resources), jsonOrValue(*expResources))
 		}
+		return false
+	}
 
-		log.Infof("Resources are not within the expected limits, updating the deployment. Actual: %+v Expected: %+v", *resources, *expResources)
-		if err := k8s.UpdateDeployment(expResources); err != nil {
-			log.Error(err)
-			continue
+	if (op == scaleDown && now.Before(lastChange.Add(scaleDownDelay))) ||
+		(op == scaleUp && now.Before(lastChange.Add(scaleUpDelay))) {
+		if log.V(3) {
+			log.V(3).Infof("Resources are not within the expected limits, Actual: %+v, accepted range: %+v. Skipping resource update because of scale up/down delay", jsonOrValue(*resources), jsonOrValue(*expResources))
```

Comment:
I actually think this should be logged unconditionally - this is important log and lack of it may introduce a lot of misunderstanding - please change to simply log.Infof

Reply:
Logging this unconditionally could spam logs with hundreds of the same messages were only the first one is really useful for the user. Therefore I have decided to change the implementation in a way that I will log this message unconditionally but only for the first time we when we are skipping the the override.